ctapipe/tools/muon_reconstruction.py

```diff
@@ -5,121 +5,223 @@
+
+        if np.count_nonzero(mask) <= self.min_pixels.tel[tel_id]:
+            self.log.debug(
+                f'Skipping event {event_id}-{tel_id}:'
+                f' Less then {self.min_pixels.tel[tel_id]} pixels on ring'
+            )
+            return
+
+        if np.isnan([ring.radius.value, ring.center_x.value, ring.center_y.value]).any():
+            self.log.debug(f'Skipping event {event_id}-{tel_id}: Ring fit did not succeed')
+            return
+
+        # add ring containment, not filled in fit
+        ring.containment = ring_containment(
+            ring.radius,
+            fov_radius,
+            ring.center_x,
+            ring.center_y,
+        )
 
-            if numev == 0:
-                _exclude_some_columns(event.inst.subarray, self.writer)
+        # intensity_fitter does not support a mask yet, set ignored pixels to 0
+        image[~mask] = 0
 
-            numev += 1
+        result = self.intensity_fitter(
+            tel_id,
+            ring.center_x,
+            ring.center_y,
+            ring.radius,
+            image,
+            pedestal=1.1,
+        )
 
-            if not muon_evt['MuonIntensityParams']:
-                # No telescopes  contained a good muon
-                continue
-            else:
-                if self.display:
-                    plot_muon_event(event, muon_evt)
+        self.log.info(
+            f'Muon fit: r={ring.radius:.2f}'
+            f', width={result.ring_width:.4f}'
+            f', efficiency={result.optical_efficiency:.2%}',
+        )
 
-                for tel_id in muon_evt['TelIds']:
-                    idx = muon_evt['TelIds'].index(tel_id)
-                    intens_params = muon_evt['MuonIntensityParams'][idx]
+        tel_event_index = TelEventIndexContainer(
+            tel_id=tel_id, tel_type_id=self.tel_type_ids[tel_id]
+        )
 
-                    if intens_params is not None:
-                        ring_params = muon_evt['MuonRingParams'][idx]
-                        cam_id = str(event.inst.subarray.tel[tel_id].camera)
-                        self.num_muons_found[cam_id] += 1
-                        self.log.debug("INTENSITY: %s", intens_params)
-                        self.log.debug("RING: %s", ring_params)
-                        self.writer.write(table_name=cam_id,
-                                          containers=[intens_params,
-                                                      ring_params])
+        self.writer.write(
+            f'dl1/event/telescope/parameters/muons',
+            [event_index, tel_event_index, ring, result]
+        )
 
-                self.log.info(
-                    "Event Number: %d, found %s muons",
-                    numev, dict(self.num_muons_found)
-                )
+    def get_fov(self, tel_id):
+        '''Guesstimate fov radius for telescope with id `tel_id`'''
+        # memoize fov calculation
+        if tel_id not in self.field_of_view:
```

Comment:
Is there a reason not to just use the `@functools.lru_cache` decorator instead of an explicit memoization?  (same for other places where you cache the result).

Reply:
I thought lru cache does not work for methods. Did I remember that wrong?

Reply:
ah, yes that is a problem. Still, making it a decorator (needs to be a custom one like mentioned here https://stackoverflow.com/questions/33672412/python-functools-lru-cache-with-class-methods-release-object, might still be clearer.

Reply:
or wait for the time we use  python >= 3.8: https://docs.python.org/3/library/functools.html#functools.cached_property